### PR TITLE
New static configuration loading system.

### LINKED ISSUE #7: Configuration through environment
Any plans to do configuration through environment? I'd love the ability to configure with something like:

``` shell
docker run \
  -e DOCKER_ENDPOINT="unix:///tmp/docker.sock" \
  -e DOCKER_DOMAIN="docker.localhost" \
  -e DOCKER_WATCH=true \
  -v /var/run/docker.sock:/tmp/docker.sock \
  emilevauge/traefik
```

### PR DESCRIPTION
**Pull request by ldez**
### What does this PR do?

New static configuration loading system that allows:

- commands and sub-commands
- flags (map and slice of struct)
- files
- environment variables

Removes Flaeg and Staert :tada: 

### Motivation

Being able to handle sub-commands as well as maps and slices of structs, and environment variables.

Since Flaeg and Staert didn't allow implementing easily the above, and given it was hard to maintain, we took the opportunity to fully replace it. :tada: 

Closes #7, #4740, #4653, #3985, #4645

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Co-authors

Co-authored-by: Mathieu Lonjaret <mathieu.lonjaret@gmail.com>

### CODE REVIEW

**Comment by ldez** [comment]
For the help, with the configuration of the v2 only display the top level is an issue:
- some options (like `acme` or `providers.docker`) cannot be enable with the root element
- to display sub-element you need to have some sub flags for flags or create a complex system to understand what element to display.

We plan to create a system to "automatically" create and update the documentation (the md files) for flags and env vars.

**Comment by ldez** [comment]
There are 2 slices of struct: `domains` and `constraints`.

For the `domains` (acme), they will be remove from the static configuration, look at #4872

For the `constraints`, they will be replace by a slice of string like in the v1.

**Comment by emilevauge** [comment]
OK for the slices topic.
About the help, I understand your points, but I really think we have to do something on this. The current implementation has become unreadable.
I was thinking of :
```
$ traefik --help
--accesslog  (Default: "false")
    Access log settings.
--acme.*
    ACME configuration
--api  (Default: "false")
    Enable api/dashboard.
--hostresolver  (Default: "false")
    Enable CNAME Flattening.

$ traefik --accesslog --help
--accesslog  (Default: "false")
    Access log settings.

--accesslog.bufferingsize  (Default: "0")
    Number of access log lines to process in a buffered way.

--accesslog.fields.defaultmode  (Default: "keep")
    Default mode for fields: keep | drop

--accesslog.fields.headers.defaultmode  (Default: "keep")
    Default mode for fields: keep | drop | redact
```

The `--acme` option is an exception indeed, but we can find a workaround.

**Comment by mpl** [comment]
> OK for the slices topic.
> About the help, I understand your points, but I really think we have to do something on this. The current implementation has become unreadable.

I do not understand why you say it has _become_ unreadable. The number of items displayed is not larger than it was before. So do you mean that:
1) it was already unreadable before this PR, because there was/are too many options ? (fair enough)
2) or you find the new format not as readable as the old format? if yes, why?

> I was thinking of :
> 
> ```
> $ traefik --help
> --accesslog  (Default: "false")
>     Access log settings.
> --acme.*
>     ACME configuration
> --api  (Default: "false")
>     Enable api/dashboard.
> --hostresolver  (Default: "false")
>     Enable CNAME Flattening.
> 
> $ traefik --accesslog --help
> --accesslog  (Default: "false")
>     Access log settings.
> 
> --accesslog.bufferingsize  (Default: "0")
>     Number of access log lines to process in a buffered way.
> 
> --accesslog.fields.defaultmode  (Default: "keep")
>     Default mode for fields: keep | drop
> 
> --accesslog.fields.headers.defaultmode  (Default: "keep")
>     Default mode for fields: keep | drop | redact
> ```
> 
> The `--acme` option is an exception indeed, but we can find a workaround.

**Comment by emilevauge** [comment]
@mpl 
>I do not understand why you say it has become unreadable. The number of items displayed is not larger than it was before. So do you mean that:
>1. it was already unreadable before this PR, because there was/are too many options ? (fair enough)
>2. or you find the new format not as readable as the old format? if yes, why?

3 years ago, it was readable, because Traefik had far less options at that time. But not anymore. So basically, it's 1/ :)

**Comment by emilevauge** [comment]
After discussing, we will work on the help issue in the near future.

File: pkg/provider/rest/rest.go [hunk 1]
```diff
@@ -19,7 +19,13 @@ var _ provider.Provider = (*Provider)(nil)
...
 type Provider struct {
 	configurationChan chan<- config.Message
-	EntryPoint        string `description:"EntryPoint" export:"true"`
+	EntryPoint        string `description:"EntryPoint." export:"true"`
+}
+
+// SetDefaults sets the default values.
+func (p *Provider) SetDefaults() {
+	p.EntryPoint = "traefik"
+	// FIXME p.EntryPoint = static.DefaultInternalEntryPointName
```
Comment (jbdoumenjou) [inline pkg/provider/rest/rest.go]: As it seems to be difficult to avoid the import cycle, is this FIXME usefull ? Same question for in types/metrics.go

**Pull request merged**